heat/core/types.py

```diff
@@ -0,0 +1,257 @@
+"""
+types: Defines the basic heat data types in the hierarchy as shown below. Design inspired by the Python package numpy.
+
+As part of the type-hierarchy: xx -- is bit-width
+generic
+ +-> bool, bool_                            (kind=?)
+ +-> number
+ |   +-> integer
+ |   |   +-> signedinteger     (intxx)      (kind=b, i)
+ |   |   |     int8, byte
+ |   |   |     int16, short
+ |   |   |     int32, int
+ |   |   |     int64, long
+ |   |   \\-> unsignedinteger  (uintxx)     (kind=B, u)
+ |   |         uint8, ubyte
+ |   \\-> floating             (floatxx)    (kind=f)
+ |         float32, float, float_
+ |         float64, double     (double)
+ \\-> flexible (currently unused, placeholder for characters)
+"""
+
+import abc
+import builtins
+import torch
+
+from .communicator import NoneCommunicator
+from . import tensor
+
+
+__all__ = [
+    'generic',
+    'number',
+    'integer',
+    'signedinteger',
+    'unsignedinteger',
+    'bool',
+    'bool_',
+    'floating',
+    'int8',
+    'byte',
+    'int16',
+    'short',
+    'int32',
+    'int',
+    'int64',
+    'long',
+    'uint8',
+    'ubyte',
+    'float32',
+    'float',
+    'float_',
+    'float64',
+    'double',
+    'flexible',
+]
+
+
+class generic(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def __new__(cls, *value):
+        try:
+            torch_type = cls.torch_type()
+        except TypeError:
+            raise TypeError('cannot create \'{}\' instances'.format(cls))
+
+        value_count = len(value)
+
+        # check whether there are too many arguments
+        if value_count >= 2:
+            raise TypeError('function takes at most 1 argument ({} given)'.format(value_count))
+        # if no value is given, we will initialize the value to be zero
+        elif value_count == 0:
+            value = ((0,),)
+
+        # otherwise, attempt to create a torch tensor of given type
+        try:
+            array = torch.tensor(*value, dtype=torch_type)
+        except TypeError as exception:
+            # re-raise the exception to be consistent with numpy's exception interface
+            raise ValueError(str(exception))
+
+        return tensor.tensor(array, tuple(array.shape), split=None, comm=NoneCommunicator())
+
+    @classmethod
+    @abc.abstractclassmethod
```

Comment:
`abc.abstractclassmethod` has become deprecated in Python 3.3, according to the [documentation ](https://docs.python.org/3/library/abc.html#abc.abstractclassmethod). `abc.abstractmethod` should be sufficient.